File: backend/src/models/rule.rs
use std::collections::BTreeMap;

use crate::helpers::{Either, MVec, null_to_default};
use serde::{Deserialize, Serialize};
use specta::Type;

use super::damage::{Damage, DamageRoll};

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(tag = "key", deny_unknown_fields)]
pub enum Rule {
    #[serde(rename_all = "camelCase")]
    ActiveEffectLike {
        mode: String,
        path: String,
        phase: Option<String>,
        value: serde_json::Value,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        priority: Option<u8>,
    },
    #[serde(rename_all = "camelCase")]
    Strike {
        fist: Option<bool>,

        damage: Option<Damage>,
        category: Option<String>,
        group: Option<String>,
        img: Option<String>,
        range: Option<serde_json::Value>,
        slug: Option<String>,
        label: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        traits: Vec<String>,
        base_type: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        options: Vec<String>,
    },
    #[serde(rename_all = "camelCase")]
    RollOption {
        count: Option<bool>,
        always_active: Option<bool>,
        mergeable: Option<bool>,
        disabled_value: Option<bool>,
        toggleable: Option<Either<String, bool>>,
        remove_upon_create: Option<bool>,

        label: Option<String>,
        option: String,
        phase: Option<String>,
        domain: Option<String>,
        placement: Option<String>,
        priority: Option<u8>,
        #[serde(default, deserialize_with = "null_to_default")]
        disabled_if: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        suboptions: Vec<RuleChoice>,
        #[serde(default, deserialize_with = "null_to_default")]
        value: Option<serde_json::Value>,
    },
    #[serde(rename_all = "camelCase")]
    BaseSpeed {
        #[serde(rename = "type")]
        speed_type: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        selector: MVec<String>,
        value: serde_json::Value, // u8 originally
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
    },
    #[serde(rename_all = "camelCase")]
    GrantItem {
        in_memory_only: Option<bool>,
        reevaluate_on_update: Option<bool>,
        allow_duplicate: Option<bool>,
        nest_feat: Option<bool>,

        preselect_coices: Option<serde_json::Value>, // lmao typo
        flag: Option<String>,
        uuid: String,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        preselect_choices: BTreeMap<String, String>,
        on_delete_actions: Option<serde_json::Value>, // Worth nothing
        alterations: Option<serde_json::Value>,       // TODO: Check
        priority: Option<u8>,
    },
    #[serde(rename_all = "camelCase")]
    ItemAlteration {
        item_id: Option<String>,
        item_type: Option<String>,
        phase: Option<String>,
        mode: String,
        label: Option<String>,
        selector: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        property: String,
        #[serde(default, deserialize_with = "null_to_default")]
        text_ref: Vec<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        value: Option<serde_json::Value>,
        priority: Option<serde_json::Value>, // Worth nothing
    },
    #[serde(rename_all = "camelCase")]
    ChoiceSet {
        adjust_name: Option<bool>,
        actor_flag: Option<bool>,

        allowed_drops: Option<ChoiceSetAllowedDrops>,
        #[serde(default, deserialize_with = "null_to_default")]
        choices: MVec<Either<RuleChoice, String>>,
        roll_option: Option<String>,
        label: Option<String>,
        selection: Option<String>,
        flag: Option<String>,
        prompt: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        priority: Option<u8>,
    },
    #[serde(rename_all = "camelCase")]
    CreatureSize {
        resize_equipment: Option<bool>,

        value: String,
        reach: Option<serde_json::Value>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
    },
    #[serde(rename_all = "camelCase")]
    FlatModifier {
        alternate: Option<bool>,
        critical: Option<bool>,
        hide_if_disabled: Option<bool>,
        from_equipment: Option<bool>,

        damage_category: Option<String>,
        damage_type: Option<String>,
        ability: Option<String>,
        label: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        slug: Option<String>,
        selector: MVec<String>,
        #[serde(rename = "type")]
        modifier_type: Option<String>,
        value: Option<serde_json::Value>,
    },
    #[serde(rename_all = "camelCase")]
    AdjustDegreeOfSuccess {
        #[serde(default)]
        adjustment: BTreeMap<String, String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        selector: MVec<String>,
        #[serde(rename = "type")]
        selector_type: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        outcome: Vec<String>,
    },
    #[serde(rename_all = "camelCase")]
    Weakness {
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(rename = "type", default, deserialize_with = "null_to_default")]
        weakness_type: MVec<String>,
        value: serde_json::Value,
    },
    #[serde(rename_all = "camelCase")]
    Immunity {
        #[serde(rename = "type")]
        immunity_type: String,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
    },
    #[serde(rename_all = "camelCase")]
    AdjustStrike {
        mode: String,
        property: String,
        value: Either<String, u8>,
        #[serde(default, deserialize_with = "null_to_default")]
        definition: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        selector: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    Sense {
        selector: MVec<String>,
        range: Option<serde_json::Value>,
        acuity: Option<String>,
        value: Option<serde_json::Value>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
    },
    #[serde(rename_all = "camelCase")]
    Resistance {
        value: serde_json::Value, // originally string
        #[serde(rename = "type")]
        #[serde(default, deserialize_with = "null_to_default")]
        resistance_type: MVec<String>,
        label: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        double_vs: Vec<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        exceptions: Vec<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        definition: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
    },
    #[serde(rename_all = "camelCase")]
    DamageDice {
        hide_if_disabled: Option<bool>,
        critical: Option<bool>,

        slug: Option<String>,
        label: Option<String>,
        category: Option<String>,
        damage_type: Option<String>,
        dice_number: Option<Either<String, u8>>,
        die_size: Option<String>,
        #[serde(rename = "override")]
        #[serde(default)]
        override_damage: BTreeMap<String, serde_json::Value>,
        selector: MVec<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        value: Option<serde_json::Value>,
    },
    #[serde(rename_all = "camelCase")]
    Note {
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        selector: MVec<String>,
        text: String,
        title: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        outcome: Vec<String>,
    },
    #[serde(rename_all = "camelCase")]
    ActorTraits {
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        add: Vec<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        remove: Vec<String>,
    },
    #[serde(rename_all = "camelCase")]
    AdjustModifier {
        suppress: Option<bool>,

        max_applications: Option<u8>,
        #[serde(rename = "type")]
        adjust_type: Option<String>,
        damage_type: Option<String>,
        relabel: Option<String>,
        mode: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        selector: MVec<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        selectors: Option<Vec<String>>,
        slug: Option<String>,
        value: Option<serde_json::Value>,
        priority: Option<u8>,
    },
    #[serde(rename_all = "camelCase")]
    TokenLight {
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(skip_serializing)]
        value: serde_json::Value, // Worth nothing
    },
    #[serde(rename_all = "camelCase")]
    Aura {
        slug: Option<serde_json::Value>,      // Worth nothing
        effects: Option<serde_json::Value>,   // Worth nothing
        radius: Option<serde_json::Value>,    // Worth nothing
        traits: Option<serde_json::Value>,    // Worth nothing
        predicate: Option<serde_json::Value>, // Worth nothing
    },
    #[serde(rename_all = "camelCase")]
    CraftingEntry {
        is_daily_prep: Option<bool>,
        is_prepared: Option<bool>,
        is_alchemical: Option<bool>,

        max_item_level: Option<serde_json::Value>,
        label: Option<String>,
        max_slots: Option<u8>,
        name: Option<String>,
        selector: String,
        #[serde(default, deserialize_with = "null_to_default")]
        craftable_items: MVec<RulePredicateFilter>,
        batch_sizes: Option<serde_json::Value>,
    },
    #[serde(rename_all = "camelCase")]
    EphemeralEffect {
        affects: Option<String>,
        uuid: String,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        selectors: MVec<String>,
    },
    #[serde(rename_all = "camelCase")]
    DamageAlteration {
        mode: String,
        slug: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        property: String,
        #[serde(default, deserialize_with = "null_to_default")]
        selectors: Vec<String>,
        value: Option<serde_json::Value>,
        priority: Option<u8>,
    },
    #[serde(rename_all = "camelCase")]
    MartialProficiency {
        label: Option<String>,
        same_as: Option<String>,
        slug: Option<String>,
        max_rank: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        definition: MVec<RulePredicateFilter>,
        value: Option<serde_json::Value>,
    },
    #[serde(rename_all = "camelCase")]
    CriticalSpecialization {
        alternate: Option<bool>,

        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        damage_dice: Option<DamageRoll>,
    },
    #[serde(rename_all = "camelCase")]
    SubstituteRoll {
        required: Option<bool>,

        selector: String,
        effect_type: Option<String>,
        label: Option<String>,
        slug: Option<String>,
        value: u8,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
    },
    #[serde(rename_all = "camelCase")]
    MultipleAttackPenalty {
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        #[serde(default, deserialize_with = "null_to_default")]
        selector: String,
        value: i8,
    },
    #[serde(rename_all = "camelCase")]
    RollTwice {
        remove_after_roll: Option<bool>,

        keep: String,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        selector: String,
    },
    #[serde(rename_all = "camelCase")]
    FastHealing {
        #[serde(rename = "type")]
        healing_type: Option<String>,
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        value: u8,
    },
    #[serde(rename_all = "camelCase")]
    DexterityModifierCap {
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        value: u8,
    },
    #[serde(rename_all = "camelCase")]
    TempHP {
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        value: u8,
    },
    #[serde(rename_all = "camelCase")]
    SpecialStatistic {
        extends: String,
        item_casting: SpecialStatisticItemCasting,
        priority: Option<u8>,
        slug: String,
    },
    #[serde(rename_all = "camelCase")]
    TokenEffectIcon {
        #[serde(default, deserialize_with = "null_to_default")]
        predicate: MVec<RulePredicateFilter>,
        value: String,
    },
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(untagged)]
pub enum RulePredicateFilter {
    String(String),
    Number(i8),
    Not {
        not: MVec<Box<RulePredicateFilter>>,
    },
    And {
        and: MVec<Box<RulePredicateFilter>>,
    },
    Or {
        or: MVec<Box<RulePredicateFilter>>,
    },
    Nand {
        nand: MVec<Box<RulePredicateFilter>>,
    },
    Nor {
        nor: MVec<Box<RulePredicateFilter>>,
    },
    Lt {
        lt: MVec<Box<RulePredicateFilter>>,
    },
    Lte {
        lte: MVec<Box<RulePredicateFilter>>,
    },
    Gt {
        gt: MVec<Box<RulePredicateFilter>>,
    },
    Gte {
        gte: MVec<Box<RulePredicateFilter>>,
    },
}

#[derive(Serialize, Deserialize, Debug, Type)]
pub struct RuleChoice {
    label: Option<String>,
    item_type: Option<String>,
    value: Option<serde_json::Value>,
    #[serde(default, deserialize_with = "null_to_default")]
    filter: MVec<RulePredicateFilter>,
}

#[derive(Serialize, Deserialize, Debug, Type)]
pub struct ChoiceSetAllowedDrops {
    label: Option<String>,
    #[serde(default, deserialize_with = "null_to_default")]
    predicate: MVec<RulePredicateFilter>,
}

#[derive(Serialize, Deserialize, Debug, Type)]
pub struct SpecialStatisticItemCasting {
    #[serde(default, deserialize_with = "null_to_default")]
    predicate: MVec<RulePredicateFilter>,
}

File: data-transformer/src/models/secondary/rule.ts
import "reflect-metadata";
import "es6-shim";
import {
    Exclude,
    Expose,
    plainToInstance,
    Transform,
    Type,
} from "class-transformer";
import { exit } from "process";
import type { Damage } from "./damage";
import { possiblyTranslate } from "../../utils/lang";

export function mapToRule(obj: any): RuleType {
    switch (obj.key) {
        case CHOICE_SET_NAME:
            return plainToInstance(ChoiceSetRule, obj);

        default:
            return obj
    }
}

export class RulePredicate { }

export type RuleType =
    | ChoiceSetRule
    | any;

export class Rule {
    key!: string;
}

const ACTIVE_EFFECT_LIKE_NAME = "ActiveEffectLike";
export class ActiveEffectLikeRule extends Rule {
    @Exclude() priority!: number;
    mode!: string;
    path!: string;
    value!: any;
}

const STRIKE_NAME = "Strike";
export class StrikeRule extends Rule {
    damage!: Damage;
    category!: string;
    group!: string;
    img!: string;
    label!: string;
    predicate!: string[];
    traits!: string[];
}

const ROLL_OPTION_NAME = "RollOption";
export class RollOptionRule extends Rule {
    alwaysActive!: boolean;
    mergeable!: boolean;
    toggleable!: boolean;

    label!: string;
    option!: string;
    predicate!: any;
    suboptions!: Map<string, RuleChoice>;
    @Exclude() priority!: any;
    @Exclude() placement!: any;
}

const BASE_SPEED_NAME = "BaseSpeed";
export class BaseSpeedRule extends Rule {
    selector!: string;
    value!: number;
}

const GRANT_ITEM_NAME = "GrantItem";
export class GrantItemRule extends Rule {
    uuid!: string;
}

const ITEM_ALTERATION_NAME = "ItemAlteration";
export class ItemAlterationRule extends Rule {
    itemType!: string;
    mode!: string;
    predicate!: string[];
    property!: string;
    @Transform(({ obj }) => {
        const value = obj.value;
        if (Array.isArray(value)) {
            const newValue = value.map((v: any) => v.text);
            Object.assign(value, newValue);
            return value;
        }
        return undefined;
    })
    @Expose()
    textRef?: string[];

    @Transform(({ obj }) => {
        const value = obj.value;
        if (Array.isArray(value)) {
            return undefined;
        }
    })
    @Expose()
    value?: any;
}

const CHOICE_SET_NAME = "ChoiceSet";
export class ChoiceSetRule extends Rule {
    adjustName!: boolean;

    @Type(() => RuleChoice)
    choices!: RuleChoice[];
    flag!: string;

    @Transform(({ obj }) => {
        if (obj.prompt) {
            const translated = possiblyTranslate(obj.prompt);
            return translated
        }
    })
    prompt!: string;

    @Transform(({ obj }) => {
        if (obj.label) {
            const translated = possiblyTranslate(obj.label);
            return translated
        }
    })
    label!: string;
}

export class RuleChoice {
    @Transform(({ obj }) => {
        if (obj.label)
            return possiblyTranslate(obj.label)
    })
    label!: string;

    @Transform(({ obj }) => {
        if (obj.prompt)
            return possiblyTranslate(obj.prompt)
    })
    prompt!: any;
    value!: any;
}

const CREATURE_SIZE_NAME = "CreatureSize";
export class CreatureSizeRule extends Rule {
    value!: string;
}

const FLAT_MODIFIER_NAME = "FlatModifier";
export class FlatModifierRule extends Rule {
    label!: string;
    predicate!: any;
    selector!: string;
    type!: string;
    value!: number;
}

const ADJUST_DEGREE_OF_SUCCESS_NAME = "AdjustDegreeOfSuccess";
export class AdjustDegreeOfSuccessRule extends Rule {
    adjustment!: {
        success: string;
    };
    predicate!: string[];
    selector!: string;
}

const WEAKNESS_NAME = "Weakness";
export class WeaknessRule extends Rule {
    predicate: any;
    type!: string;
    value!: string;
}

const IMMUNITY_NAME = "Immunity";
export class ImmunityRule extends Rule {
    type!: string;
}

const ADJUST_STRIKE_NAME = "AdjustStrike";
export class AdjustStrikeRule extends Rule {
    mode!: string;
    property!: string;
    value!: string;
    definition!: string[];
}

const SENSE_NAME = "Sense";
export class SenseRule extends Rule {
    selector!: string;
}

const RESISTANCE_NAME = "Resistance";
export class ResistanceRule extends Rule {
    type!: string;
    value!: string;
}

const DAMAGE_DICE_NAME = "DamageDice";
export class DamageDiceRule extends Rule {
    override!: {
        dieSize: string;
    };
    selector!: string;
}

const NOTE_NAME = "Note";
export class NoteRule extends Rule {
    predicate!: string[];
    selector!: string[];
    text!: string;
    title!: string;
}

const ACTOR_TRAITS_NAME = "ActorTraits";
export class ActorTraitsRule extends Rule {
    add!: string[];
    remove!: string[];
}

const ADJUST_MODIFIER_NAME = "AdjustModifier";
export class AdjustModifierRule extends Rule {
    suppress!: boolean;

    mode!: string;
    predicate!: string[];
    selector!: string;
    slug!: string;
}

const TOKEN_LIGHT_NAME = "TokenLight";
export class TokenLightRule extends Rule {
    predicate!: string[];
    value!: any;
}

const AURA_NAME = "Aura";
export class AuraRule extends Rule {
    effects!: any;
    radius!: number;
    traits!: string[];
}

const CRAFTING_ENTRY_NAME = "CraftingEntry";
export class CraftingEntryRule extends Rule {
    craftableItems!: string[];
    isDailyPrep!: boolean;
    isPrepared!: boolean;
    name!: string;
    selector!: string;
}
